platform-sdk/swirlds-state-impl/src/main/java/com/swirlds/state/merkle/MerkleStateRoot.java

```diff
@@ -293,652 +131,199 @@ public T copy() {
+                        // TODO: check possibilities for optimization
+                        InterruptableConsumer<Pair<Bytes, Bytes>> handler = pair ->
+                                virtualMap.putBytes(stateIdBytes.append(pair.key()), pair.value());
 
-            final var node = findNode(md);
-            if (node instanceof QueueNode q) {
-                final var ret = createReadableQueueState(md, q);
-                queueInstances.put(stateKey, ret);
-                return ret;
-            } else {
-                // This exception should never be thrown. Only if "findNode" found the wrong node!
-                throw new IllegalStateException("Unexpected type for queue state " + stateKey);
-            }
-        }
+                        try {
+                            logger.info(STARTUP.getMarker(), "\nMigrating {} (size: {})...", virtualMapToMigrate.getLabel(), virtualMapToMigrate.size());
+                            long migrationStartTime = System.currentTimeMillis();
+                            VirtualMapMigration.extractVirtualMapData(AdHocThreadManager.getStaticThreadManager(), virtualMapToMigrate, handler, THREAD_COUNT);
+                            long migrationTimeMs = System.currentTimeMillis() - migrationStartTime;
+                            logger.info(STARTUP.getMarker(), "Migration complete for {} took {} ms", virtualMapToMigrate.getLabel(), migrationTimeMs);
+                            logger.info(STARTUP.getMarker(), "New Virtual Map size: {}", virtualMap.size());
+                            totalMigrationTimeMs.addAndGet(migrationTimeMs);
+                            totalMigratedObjects.addAndGet(virtualMapToMigrate.size());
+                        } catch (InterruptedException e) { // TODO: revisit exception handling
+                            throw new RuntimeException(e);
+                        }
 
-        @Override
-        public boolean contains(@NonNull final String stateKey) {
-            return stateMetadata.containsKey(stateKey);
-        }
+                        if (VALIDATE_MIGRATION_FF) {
+                            long validationStartTime = System.currentTimeMillis();
+                            logger.info(STARTUP.getMarker(), "Validating the new Virtual Map contains all data from the KV State {}", virtualMapToMigrate.getLabel());
+                            validateKVStateMigrated(virtualMap, virtualMapToMigrate);
+                            long validationTimeMs = System.currentTimeMillis() - validationStartTime;
+                            logger.info(STARTUP.getMarker(), "Validation complete for the KV State {} took {} ms", virtualMapToMigrate.getLabel(), validationTimeMs);
+                            totalValidationTimeMs.addAndGet(validationTimeMs);
+                        }
+                    });
+
+            logger.info(STARTUP.getMarker(), "Migrating Queue states to the one Virtual Map...");
+
+            IntStream.range(0, getNumberOfChildren())
```

Comment:
To make it slightly easier to read, KV, queue, and singleton state migrations can be extracted into three separate methods

Reply:
I can propose to refactor it further:
- use for loop instead of streams
- extract repetitive code (logging, etc.)

Wdyt?